ta/pkcs11/src/processing_asymm.c

```diff
@@ -903,3 +933,159 @@ enum pkcs11_rc do_asymm_derivation(struct pkcs11_session *session,
+	    ctx->aes_key_bits != 256)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	mbedtls_nist_kw_init(&kw_ctx);
+	TEE_GenerateRandom(aes_key_value, aes_key_size);
+	res = TEE_AsymmetricEncrypt(proc->tee_op_handle,
+				    tee_attrs, tee_attrs_count,
+				    aes_key_value, aes_key_size,
+				    out_buf, &aes_wraped_size);
+	if (res) {
+		res = tee2pkcs_error(res);
+		goto out;
+	}
+
+	res = mbedtls_nist_kw_setkey(&kw_ctx, MBEDTLS_CIPHER_ID_AES,
+				     aes_key_value, ctx->aes_key_bits, true);
+	if (res) {
+		res = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	res = mbedtls_nist_kw_wrap(&kw_ctx, MBEDTLS_KW_MODE_KWP,
+				   data, data_sz,
+				   (uint8_t *)out_buf + aes_wraped_size,
+				   &target_key_size, *out_sz - aes_wraped_size);
+	if (res) {
+		res = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	*out_sz = target_key_size + aes_wraped_size;
+
+out:
+	mbedtls_nist_kw_free(&kw_ctx);
+	TEE_MemFill(aes_key_value, 0, aes_key_size);
+	return res;
+}
+
+static enum pkcs11_rc unwrap_rsa_aes_key(struct active_processing *proc,
+					 void *data, uint32_t data_sz,
+					 void **out_buf, uint32_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_OperationInfo info;
+	TEE_Attribute tee_attrs[2] = { };
+	struct rsa_aes_key_wrap_processing_ctx *ctx = proc->extra_ctx;
+	mbedtls_nist_kw_context kw_ctx;
+	uint8_t aes_key_value[32] = { };
+	uint32_t aes_key_size = ctx->aes_key_bits / 8;
+	uint32_t tee_attrs_count = 0;
+	uint32_t wraped_key_size = 0;
+	uint32_t rsa_key_size = 0;
+	size_t target_key_size = 0;
+
+	if (ctx->aes_key_bits != 128 &&
+	    ctx->aes_key_bits != 192 &&
+	    ctx->aes_key_bits != 256)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	TEE_GetOperationInfo(proc->tee_op_handle, &info);
+	rsa_key_size = info.keySize / 8;
+	wraped_key_size = data_sz - rsa_key_size;
+
+	*out_buf = TEE_Malloc(aes_key_size, TEE_MALLOC_FILL_ZERO);
+	if (!*out_buf)
+		return PKCS11_CKR_DEVICE_MEMORY;
+
+	mbedtls_nist_kw_init(&kw_ctx);
+	res = TEE_AsymmetricDecrypt(proc->tee_op_handle,
+				    tee_attrs, tee_attrs_count,
+				    data, rsa_key_size,
+				    aes_key_value, &aes_key_size);
+	if (res) {
+		res = tee2pkcs_error(res);
+		goto out;
+	}
+
+	res = mbedtls_nist_kw_setkey(&kw_ctx, MBEDTLS_CIPHER_ID_AES,
+				     aes_key_value, ctx->aes_key_bits, false);
+	if (res) {
+		res = PKCS11_CKR_WRAPPED_KEY_INVALID;
+		goto out;
+	}
+
+	res = mbedtls_nist_kw_unwrap(&kw_ctx, MBEDTLS_KW_MODE_KWP,
+				     (uint8_t *)data + rsa_key_size,
+				     wraped_key_size, *out_buf,
+				     &target_key_size, aes_key_size);
+	if (res)
+		res = PKCS11_CKR_WRAPPED_KEY_INVALID;
+
+	*out_sz = target_key_size;
+out:
+	if (res) {
+		TEE_Free(*out_buf);
+		*out_buf = NULL;
+		*out_sz = 0;
+	}
+
+	TEE_MemFill(aes_key_value, 0, aes_key_size);
+	mbedtls_nist_kw_free(&kw_ctx);
+	return res;
+}
+
+enum pkcs11_rc wrap_data_by_asymm_enc(struct pkcs11_session *session,
+				      void *data, uint32_t data_sz,
+				      void *out_buf, uint32_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
```

Comment:
replace straight with `enum pkcs11_rc rc = ...`

ditto in `unwrap_key_by_asymm()`

Reply:
done